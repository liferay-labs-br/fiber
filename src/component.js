import createComponent from './component-lifecycle';

class Component {
	
  /**
   * Constructor
   * isMounted prevents the component call the componentWillMount() function
   * when the user use the constructor in his Fiber component
   */
  constructor(props, context, isMounted) {
    this.props = props;
    this.context = context;

    if (isMounted) {
      this.componentWillMount();
    }
  }

  /**
   * Lifecycle
   * Corresponds created()
   */
  componentWillMount() {}

  /**
   * Lifecycle
   * Corresponds disposed()
   */
  componentWillUnmount() {}

  /**
   * Lifecycle
   * Corresponds shouldUpdate()
   */
  shouldComponentUpdate() {}

  /**
   * Lifecycle
   * Corresponds rendered()
   */
  componentDidUpdate() {}

  setState(state, callback) {}

  render() {}
}

const buildComponentFromVNode = (vnode, context) => {
  let props = Object.assign({}, vnode.attributes);
  return createComponent(vnode.nodeName, props, context);
};

export {
  Component,
  buildComponentFromVNode,
};
