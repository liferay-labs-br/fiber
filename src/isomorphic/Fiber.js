/**
 * Copyright (c) 2017-present, Fiber.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import Component from './FiberComponent';
import createElement from './FiberElement';
import renderFactory from './FiberRender';

export default {
  Component,
  createElement,
  renderFactory
};

export {
  Component,
  createElement,
  renderFactory
};
