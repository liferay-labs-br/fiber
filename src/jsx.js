export function jsx(nodeName, attributes, ...args) {
  let children = args.length ? [].concat(...args) : null;

  return { nodeName, attributes, children };
}
