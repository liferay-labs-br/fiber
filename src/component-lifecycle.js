import { Component } from './component';

export default function createComponent(Constructor, props, context) {
  let instance;

  if (Constructor.prototype && Constructor.prototype.render) {
    instance = new Constructor(props, context);
    Component.call(instance, props, context, true);
    instance = instance.render();
  } else {
    instance = Constructor(props);
  }

  return instance;
}
