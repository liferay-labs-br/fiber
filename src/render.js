import { buildComponentFromVNode } from './component.js';
import renderUtils from './utils/render-utils';

export default function render(vnode, parent) {
	if (vnode == null || typeof vnode === 'boolean') vnode = '';

	if (typeof vnode==='string') return document.createTextNode(vnode);
	
	if (typeof vnode === 'function') vnode = vnode();

	if (typeof vnode.nodeName === 'function') vnode = buildComponentFromVNode(vnode, {});
	
	let node = document.createElement(vnode.nodeName);
	
	let attributes = vnode.attributes || {};
	renderUtils._formatAttributes(node, attributes);
	
	(vnode.children || []).forEach( child => node.appendChild(render(child)) );
	
	if (parent) parent.appendChild(node);
	
	return node;
}
