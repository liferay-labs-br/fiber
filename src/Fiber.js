/**
 * Copyright (c) 2017-present, Fiber.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { Component, createElement } from './isomorphic/Fiber';
import { render } from './renderers/dom/FiberDOM';

export default {
  createElement,
  Component,
  render
}

export {
  createElement,
  Component,
  render
}
