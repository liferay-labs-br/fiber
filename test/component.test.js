import { render, createElement, Component } from '../src/Fiber';

let Bar, Foo;

describe('render', () => {
  let artboard;

  beforeAll(() => {
    artboard = document.createElement('div');
		(document.body || document.documentElement).appendChild(artboard);
  });

  beforeEach(() => {
    artboard.innerHTML = '';
  });

  afterAll(() => {
    artboard.parentNode.removeChild(artboard);
		artboard = null;

    if (Bar) {
      Bar.dispose();
    }

    if (Foo) {
      Foo.dispose();
    }
  });

  it('should not render when render of the class return empty', () => {
    class Bar extends Component {
      render() {
        return;
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when render of the class return bool false', () => {
    class Bar extends Component {
      render() {
        return false;
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when render of the class return bool true', () => {
    class Bar extends Component {
      render() {
        return true;
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when render of the class return string', () => {
    class Bar extends Component {
      render() {
        return 'Foo bar';
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render when render of the class return undefined', () => {
    class Bar extends Component {
      render() {
        return undefined;
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should render class with nested children', () => {
    class Bar extends Component {
      render() {
        return (
          <div id="App">
            <div>
              <h1>Hello World!</h1>
              <p>foo <span>bar</span></p>
            </div>
          </div>
        );
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });

  it('should not render the onClick event', () => {
    class Bar extends Component {
      click() {}

      render() {
        return (
          <div onClick={this.click} id="app">Foo bar</div>
        );
      }
    }

    render(<Bar />, artboard);

    expect(artboard).toMatchSnapshot();
  });
});
