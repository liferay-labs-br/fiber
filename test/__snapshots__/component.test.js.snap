// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should not render the onClick event 1`] = `
<div>
  <div
    id="app"
  >
    Foo bar
  </div>
</div>
`;

exports[`render should not render when render of the class return bool false 1`] = `<div />`;

exports[`render should not render when render of the class return bool true 1`] = `<div />`;

exports[`render should not render when render of the class return empty 1`] = `<div />`;

exports[`render should not render when render of the class return string 1`] = `<div />`;

exports[`render should not render when render of the class return undefined 1`] = `<div />`;

exports[`render should render class with nested children 1`] = `
<div>
  <div
    id="App"
  >
    <div>
      <h1>
        Hello World!
      </h1>
      <p>
        foo 
        <span>
          bar
        </span>
      </p>
    </div>
  </div>
</div>
`;
